01-getting-started/projects/final/app/src/main/java/com/raywenderlich/podplay/ui/PodcastActivity.kt

```diff
@@ -0,0 +1,312 @@
+import com.raywenderlich.podplay.repository.ItunesRepo
+import com.raywenderlich.podplay.repository.PodcastRepo
+import com.raywenderlich.podplay.service.ItunesService
+import com.raywenderlich.podplay.service.RssFeedService
+import com.raywenderlich.podplay.ui.PodcastDetailsFragment.OnPodcastDetailsListener
+import com.raywenderlich.podplay.viewmodel.PodcastViewModel
+import com.raywenderlich.podplay.viewmodel.SearchViewModel
+import com.raywenderlich.podplay.worker.EpisodeUpdateWorker
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import java.util.concurrent.TimeUnit
+
+class PodcastActivity : AppCompatActivity(), PodcastListAdapterListener,
+    OnPodcastDetailsListener {
+
+  private val searchViewModel by viewModels<SearchViewModel>()
+  private val podcastViewModel by viewModels<PodcastViewModel>()
+  private lateinit var podcastListAdapter: PodcastListAdapter
+  private lateinit var searchMenuItem: MenuItem
+  private lateinit var databinding: ActivityPodcastBinding
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    databinding = ActivityPodcastBinding.inflate(layoutInflater)
+    setContentView(databinding.root)
+    setupToolbar()
+    setupViewModels()
+    updateControls()
+    setupPodcastListView()
+    handleIntent(intent)
+    addBackStackListener()
+    scheduleJobs()
+  }
+
+  override fun onCreateOptionsMenu(menu: Menu): Boolean {
+    val inflater = menuInflater
+    inflater.inflate(R.menu.menu_search, menu)
+
+    searchMenuItem = menu.findItem(R.id.search_item)
+    val searchView = searchMenuItem.actionView as SearchView
+
+    searchMenuItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
+      override fun onMenuItemActionExpand(p0: MenuItem?): Boolean {
+        databinding.emptyListViewGroup.visibility = View.GONE
+        return true
+      }
+
+      override fun onMenuItemActionCollapse(p0: MenuItem?): Boolean {
+        showSubscribedPodcasts()
+        return true
+      }
+    })
+
+    val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
+    searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
+
+    if (supportFragmentManager.backStackEntryCount > 0) {
+      databinding.podcastRecyclerView.visibility = View.INVISIBLE
+    }
+
+    if (databinding.podcastRecyclerView.visibility == View.INVISIBLE) {
+      searchMenuItem.isVisible = false
+    }
+
+    return true
+  }
+
+  override fun onNewIntent(intent: Intent) {
+    super.onNewIntent(intent)
+    // TODO 1
+    setIntent(intent)
+    handleIntent(intent)
+  }
+
+
+  override fun onShowDetails(podcastSummaryViewData: SearchViewModel.PodcastSummaryViewData) {
+    podcastSummaryViewData.feedUrl ?: return
+    showProgressBar()
+    // TODO 2
+    podcastViewModel.viewModelScope.launch(context = Dispatchers.Main) {
```

Comment:
😁😁😁😁
Please, before anyone sees this, refactor this to call the method from VM and run the coroutine in it. After the action is done, **post** a Livedata  value which will trigger hiding progress bar and fragment switching.

Dispatchers.Main is redundant here, this is run on Main by default

Reply:
We can ommit the 'Dispatchers.Main' param, but as suggested earlier - the whole Coroutines & ViewModel scope needs to be improvised if we want to do that properly - I'm afraid it'll cost us big time!